File: samples/perl/retrieve-cui-or-code.pl
#!/usr/bin/perl

## Compatible with UMLS REST API - version 0.60 Beta
## This file retrieves basic information of a source-asserted identifer or UMLS CUI.
## usage: perl retrieve-concept-info.pl -u your-umls-username -p your-umls-password -v version -i identifer -s source vocabulary
## If you do not provide the version parameter the script queries the latest avaialble UMLS publication.
## To query the UMLS, omit the 'source' parameter and use UMLS CUI as your 'identifier' parameter.
## The full list of query parameters and fields available for CUI retrieval is at:
## https://documentation.uts.nlm.nih.gov/rest/concept/index.html and
## https://documentation.uts.nlm.nih.gov/rest/source-asserted-identifiers/index.html

use lib "lib";
use strict;
use warnings;
use URI;
use Authentication::TicketClient;
use JSON;
use REST::Client;
use Data::Dumper;
use Getopt::Std;
our ($opt_u,$opt_p,$opt_v,$opt_s, $opt_i);
getopt('upvsi');
my $username = $opt_u || die "please provide username";
my $password = $opt_p || die "please provide password";
my $id = $opt_i || die "please provide a UMLS CUI or source-asserted identifier";
my $version = defined $opt_v ? $opt_v : "current";
my $source = $opt_s;


##create a ticket granting ticket for the session
my $ticketClient = new TicketClient(username=>$opt_u,password=>$opt_p,service=>"http://umlsks.nlm.nih.gov",tgt=>"") || die "could not create TicketClient() object";
my $tgt = $ticketClient->getTgt();
my $uri = new URI("https://uts-ws.nlm.nih.gov");
my %parameters = ();
my $client = REST::Client->new();


    my $path = defined $source ? "/rest/content/".$version."/source/".$source."/".$id : "/rest/content/".$version."/CUI/".$id;
    my $json = run_query($path,\%parameters);
    my $cuiOrCode = $json->{result};

    printf "%s\n","ui- ".$cuiOrCode->{ui};
    printf "%s\n","Number of Atoms- ".$cuiOrCode->{atomCount};
    printf "%s\n","Name- ".$cuiOrCode->{name};
    printf "%s\n","Atoms- ".$cuiOrCode->{atoms};
    printf "%s\n","Definitions- ".$cuiOrCode->{definitions};
    printf "%s\n","Relations- ".$cuiOrCode->{relations};
    printf "%s\n","Highest Ranking Atom- ".$cuiOrCode->{defaultPreferredAtom};
    my $stys = $cuiOrCode->{semanticTypes};
    
    ## Semantic Types are only assigned to UMLS CUIs. There are often more than one semantic type per CUI.
    if (!defined $source) {
       
       printf "%s\n", "Semantic Types:";
       foreach my $sty (@{ $stys }) {
	  printf "%s\n", "uri: ".$sty->{uri};
	  printf "%s\n", "name: ".$sty->{name};
	
        }
    }  
    ##source-asserted data. UMLS CUIs do not have transitive relations.
    printf "%s\n","Parents- ".$cuiOrCode->{parents} if defined $source;
    printf "%s\n","Children- ".$cuiOrCode->{children} if defined $source;
    printf "%s\n","Ancestors- ".$cuiOrCode->{ancestors} if defined $source;
    printf "%s\n","Descendants- ".$cuiOrCode->{descendants} if defined $source;
    
   

sub format_json {
	my $json_in = shift;
	my $json = JSON->new;
	my $obj = $json->decode($json_in);
	#print Dumper(\$obj);
	return $obj;
}

sub run_query {
	
	my ($path, $parameters) = @_;	
	$parameters{ticket} = $ticketClient->getServiceTicket();
	$uri->path($path);
	$uri->query_form($parameters);
	print qq{$uri\n};
	my $query = $client->GET($uri) || die "Could not execute query$!";
	my $results = $query->responseCode() eq '200'? $query->responseContent: die "Could not execute query$!";
	my $json = format_json($results);
	return $json;
}

File: samples/python/retrieve-cui-or-code.py
#################################################################################################
# usage of the script
# usage: python retrieve-cui-or-code.py -u USERNAME -p PASSWORD -v VERSION -i IDENTIFIER -s SOURCE
# If you do not provide the -s parameter, the script assumes you are retrieving information for a
# known UMLS CUI
#################################################################################################

from Authentication import *
import requests
import json
import argparse

parser = argparse.ArgumentParser(description='process user given parameters')
parser.add_argument("-u", "--username", required =  True, dest="username", help = "enter username")
parser.add_argument("-p", "--password", required =  True, dest="password", help = "enter passowrd")
parser.add_argument("-v", "--version", required =  False, dest="version", default = "current", help = "enter version example-2015AA")
parser.add_argument("-i", "--identifier", required =  True, dest="identifier", help = "enter identifier example-C0018787")
parser.add_argument("-s", "--source", required =  False, dest="source", help = "enter source name if known")

args = parser.parse_args()

username = args.username
password = args.password
version = args.version
identifier = args.identifier
source = args.source
AuthClient = Authentication(username,password)

###################################
#get TGT for our session
###################################

tgt = AuthClient.gettgt()
uri = "https://uts-ws.nlm.nih.gov"

try:
   source
except NameError:
   source = None

##if we don't specify a source vocabulary, assume we're retrieving UMLS CUIs
if source is None:
    content_endpoint = "/rest/content/"+str(version)+"/CUI/"+str(identifier)

else:
    content_endpoint = "/rest/content/"+str(version)+"/source/"+str(source)+"/"+str(identifier)

##ticket is the only parameter needed for this call - paging does not come into play because we're only asking for one Json object
query = {'ticket':AuthClient.getst(tgt)}
r = requests.get(uri+content_endpoint,params=query)
r.encoding = 'utf-8'
items  = json.loads(r.text)
jsonData = items["result"]

##uncomment the print statment if you want the raw json output, or you can just look at the documentation :=)
#https://documentation.uts.nlm.nih.gov/rest/concept/index.html#sample-output
#https://documentation.uts.nlm.nih.gov/rest/source-asserted-identifiers/index.html#sample-output
#print (json.dumps(items, indent = 4))

############################
### Print out fields ####

classType = jsonData["classType"]
name = jsonData["name"]
ui = jsonData["ui"]
AtomCount = jsonData["atomCount"]
Definitions = jsonData["definitions"]
Atoms = jsonData["atoms"]
DefaultPreferredAtom = jsonData["defaultPreferredAtom"]

## print out the shared data elements that are common to both the 'Concept' and 'SourceAtomCluster' class
print ("classType: " + classType)
print ("ui: " + ui)
print ("Name: " + name)
print ("AtomCount: " + str(AtomCount))
print ("Atoms: " + Atoms)
print ("Default Preferred Atom: " + DefaultPreferredAtom)

## These data elements may or may not exist depending on what class ('Concept' or 'SourceAtomCluster') you're dealing with so we check for each one.
try:
   jsonData["definitions"]
   print ("definitions: " + jsonData["definitions"])
except:
      pass

try:
   jsonData["parents"]
   print ("parents: " + jsonData["parents"])
except:
      pass

try:
   jsonData["children"]
   print ("children: " + jsonData["children"])
except:
      pass

try:
   jsonData["relations"]
   print ("relations: " + jsonData["relations"])
except:
      pass

try:
   jsonData["descendants"]
   print ("descendants: " + jsonData["descendants"])
except:
      pass

try:
   jsonData["semanticTypes"]
   print("Semantic Types:")
   for stys in jsonData["semanticTypes"]:
       print("uri: "+ stys["uri"])
       print("name: "+ stys["name"])
      
except:
      pass
